File: frontend/src/features/email/__tests__/cleanup-filter.test.mjs
import { strict as assert } from 'node:assert'
import { test } from 'node:test'
import {
  emailDateToMs,
  formatEmailRelTime,
  hasCleanupConstraint,
  matchCleanup,
} from '../cleanup-filter.ts'

test('account-only filter is not a batch-delete constraint', () => {
  assert.equal(hasCleanupConstraint({}), false)
  assert.equal(hasCleanupConstraint({ accountId: 'acct-1' }), false)
  assert.equal(hasCleanupConstraint({ subject: '发票' }), true)
  assert.equal(hasCleanupConstraint({ from: 'promo@' }), true)
  assert.equal(hasCleanupConstraint({ since: 100 }), true)
  assert.equal(hasCleanupConstraint({ until: 200 }), true)
})

test('matchCleanup uses subject, source and unix-second range', () => {
  const e = {
    fromAddress: 'promo@shop.com',
    fromName: 'Shop Promo',
    subject: '限时抢购 全场秒杀',
    date: 1_725_000_000,
  }
  assert.equal(matchCleanup(e, { subject: '抢购' }), true)
  assert.equal(matchCleanup(e, { subject: '会议纪要' }), false)
  assert.equal(matchCleanup(e, { from: 'shop.com' }), true)
  assert.equal(matchCleanup(e, { from: 'Promo' }), true)
  assert.equal(matchCleanup(e, { since: 1_724_000_000, until: 1_726_000_000 }), true)
  assert.equal(matchCleanup(e, { since: 1_726_000_001 }), false)
})

test('emailDateToMs treats unix seconds as seconds', () => {
  assert.equal(emailDateToMs(1_725_000_000), 1_725_000_000_000)
  assert.equal(emailDateToMs(1_725_000_000_000), 1_725_000_000_000)
})

test('formatEmailRelTime does not treat seconds as milliseconds', () => {
  const nowSec = Math.floor(Date.now() / 1000)
  const label = formatEmailRelTime(nowSec - 3600)
  assert.match(label, /小时前/)
  assert.equal(label.includes('20683'), false)
})

File: backend/internal/email/cleanup_filter_test.go
package email

import "testing"

func TestCleanupFilterRequiresConstraint(t *testing.T) {
	empty := CleanupFilter{}
	if empty.HasConstraint() {
		t.Fatal("empty filter must not count as a constraint")
	}
	if err := empty.Validate(); err == nil {
		t.Fatal("empty filter must fail validation")
	}
	if !(CleanupFilter{Subject: "发票"}).HasConstraint() {
		t.Fatal("subject is a constraint")
	}
	if !(CleanupFilter{From: "promo@"}).HasConstraint() {
		t.Fatal("from is a constraint")
	}
	if !(CleanupFilter{Since: 100}).HasConstraint() {
		t.Fatal("since is a constraint")
	}
	if !(CleanupFilter{Until: 200}).HasConstraint() {
		t.Fatal("until is a constraint")
	}
	if (CleanupFilter{AccountID: "acct-1"}).HasConstraint() {
		t.Fatal("account alone is not enough to batch-delete")
	}
}

func TestMatchCleanupSubjectFromDate(t *testing.T) {
	e := Email{
		FromAddress: "promo@shop.com",
		FromName:    "Shop Promo",
		Subject:     "限时抢购 全场秒杀",
		Date:        1_725_000_000,
	}
	if !MatchCleanup(e, CleanupFilter{Subject: "抢购"}) {
		t.Fatal("subject contains should match")
	}
	if MatchCleanup(e, CleanupFilter{Subject: "会议纪要"}) {
		t.Fatal("unrelated subject must not match")
	}
	if !MatchCleanup(e, CleanupFilter{From: "shop.com"}) {
		t.Fatal("from address contains should match")
	}
	if !MatchCleanup(e, CleanupFilter{From: "promo"}) {
		t.Fatal("from name contains should match")
	}
	if !MatchCleanup(e, CleanupFilter{Since: 1_724_000_000, Until: 1_726_000_000}) {
		t.Fatal("date in range should match")
	}
	if MatchCleanup(e, CleanupFilter{Since: 1_726_000_001}) {
		t.Fatal("date before since must not match")
	}
	if MatchCleanup(e, CleanupFilter{Until: 1_724_000_000}) {
		t.Fatal("date after until must not match")
	}
}

func TestSelectDeletableOnlyMovedUIDs(t *testing.T) {
	items := []CleanupItem{
		{ID: "a", AccountID: "acct-1", UID: 11},
		{ID: "b", AccountID: "acct-1", UID: 12},
		{ID: "c", AccountID: "acct-2", UID: 21},
		{ID: "d", AccountID: "acct-1", UID: 0},
	}
	moved := map[string][]int64{"acct-1": {12}}
	got := SelectDeletable(items, moved)
	if len(got) != 1 || got[0] != "b" {
		t.Fatalf("deletable=%v, want [b]", got)
	}
}
